hspec-core/src/Test/Hspec/Core/Formatters/V2.hs

```diff
@@ -258,15 +260,34 @@ defaultFailedFormatter = do
     writeLine "Failures:"
     writeLine ""
 
-    forM_ (zip [1..] failures) $ \x -> do
+    computeDiffs <- useDiff
+    precomputedDiffs <-
+      if computeDiffs
+        then do
+          pretty <- prettyPrintFunction
+          liftIO $ mapConcurrently (precomputeDiff pretty) failures
+        else return $ repeat Nothing
+
+    forM_ (zip3 [1..] failures precomputedDiffs) $ \x -> do
       formatFailure x
       writeLine ""
 
     write "Randomized with seed " >> usedSeed >>= writeLine . show
     writeLine ""
   where
-    formatFailure :: (Int, FailureRecord) -> FormatM ()
-    formatFailure (n, FailureRecord mLoc path reason) = do
+    precomputeDiff :: Maybe PrettyPrintFunction -> FailureRecord -> IO (Maybe [Diff])
+    precomputeDiff pretty (FailureRecord _ _ reason) =
+      case reason of
+        ExpectedButGot _ expected_ actual_ -> do
+          let
+            threshold = 2 :: Seconds
+            (expected, actual) = case pretty of
+              Just f -> f expected_ actual_
+              Nothing -> (expected_, actual_)
+          timeout threshold (evaluate $ diff expected actual)
```

Comment:
I actually don’t see where we would sequence the list that is returned by `diff expected actual`. @blackheaven can you try to sequence the list and possibly also make the `Diff` data type strict? Does this give you better numbers?

Also, when you first experimented with this. did you get full utilization of your 20 cores?

Reply:
> Also, when you first experimented with this. did you get full utilization of your 20 cores?

Looking at your numbers, I think you didn’t. I think we are not sequencing. Right now I’m puzzled how the `timeout` could possibly work at all.